add box mode convert in box_utils

Can-Zhao opened the pull request:
Signed-off-by: Can Zhao <canz@nvidia.com>

Fixes #3571  .

### Description
Add box mode converters into box_utils 

### Status
**Ready**

### Types of changes
<!--- Put an `x` in all the boxes that apply, and remove the not applicable items -->
- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [x] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

Nic-Ma commented:
Hi @Can-Zhao ,

Thanks for putting together the first PR for the detection task.
I tried to review the PR but I feel it's too big(more than 1000 lines) to put detailed comments, could you please help split it into smaller PRs? Then we can discuss more details in the code line.
And I would suggest to add some small examples in the doc-string of several complicated box utilities?
Especially for users like me who didn't participate in your previous design & development discussions.
For example, we tried to help users understand the bundle APIs with examples:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/bundle/config_parser.py#L407
https://github.com/Project-MONAI/MONAI/blob/dev/monai/bundle/config_parser.py#L37

What do you think?
CC @wyli @dongyang0122 

Thanks in advance.

Can-Zhao commented:
> Hi @Can-Zhao ,
> 
> Thanks for putting together the first PR for the detection task. I tried to review the PR but I feel it's too big(more than 1000 lines) to put detailed comments, could you please help split it into smaller PRs? Then we can discuss more details in the code line. And I would suggest to add some small examples in the doc-string of several complicated box utilities? Especially for users like me who didn't participate in your previous design & development discussions. For example, we tried to help users understand the bundle APIs with examples: https://github.com/Project-MONAI/MONAI/blob/dev/monai/bundle/config_parser.py#L407 https://github.com/Project-MONAI/MONAI/blob/dev/monai/bundle/config_parser.py#L37
> 
> What do you think? CC @wyli @dongyang0122
> 
> Thanks in advance.

Thank you so much Nic! I will split it into small PRs!

Can-Zhao commented:
> Hi @Can-Zhao ,
> 
> Thanks for putting together the first PR for the detection task. I tried to review the PR but I feel it's too big(more than 1000 lines) to put detailed comments, could you please help split it into smaller PRs? Then we can discuss more details in the code line. And I would suggest to add some small examples in the doc-string of several complicated box utilities? Especially for users like me who didn't participate in your previous design & development discussions. For example, we tried to help users understand the bundle APIs with examples: https://github.com/Project-MONAI/MONAI/blob/dev/monai/bundle/config_parser.py#L407 https://github.com/Project-MONAI/MONAI/blob/dev/monai/bundle/config_parser.py#L37
> 
> What do you think? CC @wyli @dongyang0122
> 
> Thanks in advance.

Hi Nic. I think it should be ready for review. Thank you so much!

Nic-Ma commented:
/black

Nic-Ma reviewed (commented):
Thanks for putting together the utilities.
Put some minor comments inline.
I see several mappings of `box mode` and `spatial dims` in the code logic, like: `int(bbox.shape[1] / 2)`, `int(len(mode) / 2)`
Maybe we should define it easier and clearer to use?
For example: associate properties `bounding box`, `mode`, `spatial dims`, etc. as a **class** and define several class functions to do the mode conversion?
@wyli @dongyang0122 What do you think?

Thanks in advance.

Can-Zhao commented:
> Thanks for putting together the utilities. Put some minor comments inline. I see several mappings of `box mode` and `spatial dims` in the code logic, like: `int(bbox.shape[1] / 2)`, `int(len(mode) / 2)` Maybe we should define it easier and clearer to use? For example: associate properties `bounding box`, `mode`, `spatial dims`, etc. as a **class** and define several class functions to do the mode conversion? @wyli @dongyang0122 What do you think?
> 
> Thanks in advance.

Thank you Nic. We discussed it before and decided not to put them into a class, as it requires the users to understand a new data structure

Can-Zhao commented on monai/data/box_utils.py:
```diff
@@ -0,0 +1,360 @@
+# TO_REMOVE = 1  if in 'xxyy','xxyyzz' mode, the bottom-right corner is included in the box,
+#       i.e., when xmin=1, xmax=2, we have w = 2
+# Currently only TO_REMOVE = 0 has been tested. Please use TO_REMOVE = 0
+TO_REMOVE = 0  # xmax-xmin = w -TO_REMOVE.
```
As mentioned in the comment, it is set to be fixed value for now. So we would not want the users to change it currently. In the future, we would test TO_REMOVE = 1 and let the user to change it.

Nic-Ma replied:
OK, I see.
Could you please help slightly adjust the comments? because usually, we don't suppose users to use MONAI source code in their project, instead, they just install MONAI and call your APIs. Comments can be:
```
Currently, only `TO_REMOVE = 0` is supported
```
What do you think?

Thanks.

Nic-Ma commented:
> > Thanks for putting together the utilities. Put some minor comments inline. I see several mappings of `box mode` and `spatial dims` in the code logic, like: `int(bbox.shape[1] / 2)`, `int(len(mode) / 2)` Maybe we should define it easier and clearer to use? For example: associate properties `bounding box`, `mode`, `spatial dims`, etc. as a **class** and define several class functions to do the mode conversion? @wyli @dongyang0122 What do you think?
> > Thanks in advance.
> 
> Thank you Nic. We discussed it before and decided not to put them into a class, as it requires the users to understand a new data structure. This is adopted by torchvision https://pytorch.org/vision/stable/_modules/torchvision/ops/boxes.html

Hi @wyli @Can-Zhao ,

OK, if you don't want to manage bounding box in a class, can we try to define the `mode` as a class? It can also solve the mode name and Enum problem, easier for future extension, some draft code for the idea:
```py
class BoxMode:
    def __init__(self, spatial_dims: int):
        self.spatial_dim = spatial_dims

    @abstractmethod
    def box_to_corner(self, box: torch.Tensor) -> Tuple:
        raise NotImplementedError(f"Subclass {self.__class__.__name__} must implement this method.")

    @abstractmethod
    def corner_to_box(self, corner: Sequence) -> torch.Tensor:
        raise NotImplementedError(f"Subclass {self.__class__.__name__} must implement this method.")


class CornerCornerMode(BoxMode):
    def box_to_corner(self, box: torch.Tensor) -> Tuple:
        if self.spatial_dim == 3:
            xmin, ymin, zmin, xmax, ymax, zmax = bbox.split(1, dim=-1)
            return xmin, xmax, ymin, ymax, zmin, zmax
        else:
            xmin, ymin, xmax, ymax = bbox.split(1, dim=-1)
            return xmin, xmax, ymin, ymax

    def corner_to_box(self, c: Sequence) -> torch.Tensor:
        if self.spatial_dim == 3:
            return torch.cat((c[0], c[2], c[4], c[1], c[3], c[5]), dim=-1)
        else:
            return torch.cat((c[0], c[2], c[1], c[3]), dim=-1)


class XxYxZxMode(BoxMode):
    def box_to_corner(self, box: torch.Tensor) -> Tuple:
        ...

    def corner_to_box(self, c: Sequence) -> torch.Tensor:
        ...


class CornerSizeMode(BoxMode):
    def box_to_corner(self, box: torch.Tensor) -> Tuple:
        ...

    def corner_to_box(self, c: Sequence) -> torch.Tensor:
        ...


class CenterSizeMode(BoxMode):
    def box_to_corner(self, box: torch.Tensor) -> Tuple:
        ...

    def corner_to_box(self, c: Sequence) -> torch.Tensor:
        ...


def box_convert_mode(bbox: NdarrayOrTensor, from_mode: BoxMode, to_mode: Optional[BoxMode]):
    if to_mode is None:
        to_mode = CornerCornerMode(spatial_dims=get_dimenstions(bbox))
    bbox_t, *_ = convert_data_type(bbox, torch.Tensor)
    corners = from_mode.box_to_corner(bbox_t)
    return to_mode.corner_to_box(corners)
```

What do you think?

Thanks.

Nic-Ma commented on monai/data/box_mode.py:
```diff
@@ -0,0 +1,331 @@
+        elif spatial_dims == 2:
+            xmin, ymin, xmax, ymax = boxes.split(1, dim=-1)
+            corner = xmin, ymin, xmax, ymax
+        if self.check_corner(corner):
```
As the above comment, if put this `check_corner` function in the `box_utils`, I feel no need to check corner in every subclass of `BoxMode`.
How about just returning the converted corners (same as boxes, BTW seems you don't check valid box format?), and do the check in the outside logic?
What do you think?

Thanks.

Nic-Ma commented on monai/data/box_utils.py:
```diff
@@ -0,0 +1,173 @@
+StandardMode = CornerCornerModeTypeA
+
+
+def get_boxmode(mode: Union[str, BoxMode, None] = None) -> BoxMode:
```
This `mapping` logic is duplicated to the `name` variable in every `BoxMode` class.
I think maybe we can change this overall logic base on my above comments:
```py
SUPPORTED_SPATIAL_DIMS = [2, 3]

SUPPORTED_MODES: [
    CornerCornerModeTypeA,
    CornerCornerModeTypeB,
    CornerCornerModeTypeC,
    CornerSizeMode,
    CenterSizeMode,
]

def get_boxmode(mode: Union[str, BoxMode, Type[BoxMode], None] = None, *args, **kwargs) -> BoxMode:
    if isinstance(mode, BoxMode):
        return mode
    if issublcass(mode, BoxMode):
        return mode(*args, **kwargs)
    elif isinstance(mode, str):
        for m in SUPPORTED_MODES:
            for n in SUPPORTED_SPATIAL_DIMS:
                if m.get_name(n) == mode
                    return m(*args, **kwargs)
    elif mode is None:
        return StandardMode(*args, **kwargs)
    raise ValueError(f"unsupported box mode: {mode}.")
```

What do you think?

Thanks.

Nic-Ma commented on monai/utils/enums.py:
```diff
@@ -311,3 +311,15 @@ class JITMetadataKeys(Enum):
     DESCRIPTION = "description"
+
+
+class BoundingBoxMode:
```
May I know why not define the class as a subclass of `Enum`?

Thanks.

Nic-Ma replied:
`BoundingBoxMode.XYXY` is not a string. You can find all the python Enum APIs in this page: https://docs.python.org/3/library/enum.html.

Thanks.

Can-Zhao commented on monai/data/box_mode.py:
```diff
@@ -0,0 +1,331 @@
+# TO_REMOVE = 1  if the bottom-right corner pixel/voxel is included in the box,
+#       i.e., when xmin=1, xmax=2, we have w = 2
+# Currently, only `TO_REMOVE = 0.` is supported
+TO_REMOVE = 0.0  # xmax-xmin = w -TO_REMOVE.
```
bounding box could be float. I will change the comment

Nic-Ma commented on monai/data/box_utils.py:
```diff
@@ -0,0 +1,291 @@
+        box_error = box_error | (corners[spatial_dims + axis] < corners[axis])
+    if box_error.sum() > 0:
+        return False
+    else:
```
This `else` will fail CI format test, maybe we can simplify it to:
```py
spatial_dims = get_dimension(corners=corners)
for axis in range(spatial_dims):
    if corners[spatial_dims + axis] < corners[axis]
        return False
return True
```
What do you think?

Thanks.

Nic-Ma commented on monai/data/box_utils.py:
```diff
@@ -0,0 +1,291 @@
+                "xyzwhd": boxes has format [xmin, ymin, zmin, xsize, ysize, zsize]
+                "ccwh": boxes has format [xcenter, ycenter, xsize, ysize]
+                "cccwhd": boxes has format [xcenter, ycenter, zcenter, xsize, ysize, zsize]
+            #. BoxMode class: choose from
```
Same comment as above.

Thanks.

Nic-Ma commented on monai/data/box_utils.py:
```diff
@@ -0,0 +1,291 @@
+    if type(src_boxmode) is type(dst_boxmode):
+        return deepcopy(boxes)
+    # convert mode
+    else:
```
No need this `else` line.

Thanks.

Nic-Ma commented on monai/data/box_utils.py:
```diff
@@ -0,0 +1,291 @@
+    src_boxmode = get_boxmode(src_mode)
+    dst_boxmode = get_boxmode(dst_mode)
+    if type(src_boxmode) is type(dst_boxmode):
+        return deepcopy(boxes)
```
May I know why `deepcopy` here?

Thanks.

Can-Zhao replied:
I was trying to make sure the returned boxes does not share memory with input boxes

Nic-Ma replied:
OK, I just feel it conflicts with your above comment: "if mode not changed, return original boxes", maybe you can change the comment to "if mode not changed, deepcopy the original boxes".

Thanks.

Nic-Ma commented on monai/data/box_utils.py:
```diff
@@ -0,0 +1,291 @@
+
+
+def convert_box_to_standard_mode(
+    boxes: NdarrayOrTensor, mode: Union[str, BoxMode, Type[BoxMode], None] = None
```
Here we must try to keep `boxes` and `mode` in-sync, that's why I proposed to define a `class` for the bounding box with mode property. I am OK for the current logic, let's keep it in mind for the future enhancement.

Thanks.

Can-Zhao replied:
Yes, I understand. We could discuss it in the future

Nic-Ma commented on monai/utils/enums.py:
```diff
@@ -311,3 +311,15 @@ class JITMetadataKeys(Enum):
     DESCRIPTION = "description"
+
+
+class BoundingBoxMode(Enum):
```
I feel this `BoundingBoxMode` is easy to be confused with `BoxMode` class.
Can we change it to `BoxModeName` or some other better class name?

Thanks.

Nic-Ma commented on tests/test_box_utils.py:
```diff
@@ -0,0 +1,135 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# import random
```
Please remove this unused line.

Thanks.

Nic-Ma commented on tests/test_box_utils.py:
```diff
@@ -0,0 +1,135 @@
+
+import numpy as np
+
+# import torch
```
Please remove this unused line.

Thanks.

Nic-Ma commented on tests/test_box_utils.py:
```diff
@@ -0,0 +1,135 @@
+
+# box_affine, box_area, box_center, box_center_dist, box_clip_to_image,
+# box_giou, box_interp, box_iou, box_pair_giou, center_in_boxes,
+# convert_to_list, non_max_suppression, resize_boxes,
```
Please remove these unused lines.

Thanks.

Nic-Ma commented on tests/test_box_utils.py:
```diff
@@ -0,0 +1,135 @@
+        expected_box = convert_data_type(expected_box, dtype=np.float32)[0]
+        boxes1 = convert_data_type(input_data["boxes"], dtype=np.float32)[0]
+        mode1 = input_data["mode"]
+        # spatial_size = input_data["spatial_size"]
```
Please remove this unused line.

Thanks.

Nic-Ma commented on tests/test_box_utils.py:
```diff
@@ -0,0 +1,135 @@
+# import torch
+from parameterized import parameterized
+
+from monai.data.box_mode import CornerCornerModeTypeA, CornerCornerModeTypeB, CornerSizeMode
```
Could you please help enhance the test cases to cover at least all the supported modes?

Thanks.

Nic-Ma reviewed (commented):
Hi @Can-Zhao ,

Thanks for your quick update.
The overall design look good to me now, put some minor comments inline.

Thanks.

Nic-Ma commented:
Hi @Can-Zhao ,

BTW, please add these new components to the `monai/data/__init__.py` file.
And update the documentation file for them:
https://github.com/Project-MONAI/MONAI/blob/dev/docs/source/data.rst
You can refer to the `ImageWriter` module:
https://github.com/Project-MONAI/MONAI/blob/dev/docs/source/data.rst#image-writer

Thanks.

Can-Zhao commented on monai/data/box_mode.py:
```diff
@@ -0,0 +1,266 @@
+            return xmin, ymin, zmin, xmax, ymax, zmax
+        elif spatial_dims == 2:
+            xmin, xmax, ymin, ymax = boxes.split(1, dim=-1)
+            corners = xmin, ymin, xmax, ymax
```
sorry. I was writing this weird code to pass mypy...

Can-Zhao reviewed (commented):
I have a question. Many functions of box_mode and box_utils for now are used only internally. Should I still add them into __init__.py ?

Can-Zhao commented:
> monai/data/__init__.py

I have a question. Many functions of box_mode and box_utils for now are used only internally. Should I still add them into monai/data/__init__.py ?

Nic-Ma reviewed (approved):
Hi @Can-Zhao ,

Thanks for your quick update.
It overall looks good to me now, just some minor comments:
1. Please add your new `Enum` and `convert_to_list` to:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/utils/__init__.py
2. Please add these new components to the docs.
3. Ensure all the CI tests can pass.

@wyli Could you please also help take a look? Do you have any other comments?

Thanks.

Can-Zhao commented:
> Hi @Can-Zhao ,
> 
> Thanks for your quick update. It overall looks good to me now, just some minor comments left:
> 
> 1. Please add your new `Enum` and `convert_to_list` to:
>    https://github.com/Project-MONAI/MONAI/blob/dev/monai/utils/__init__.py
> 2. Please add these new components to the docs.
> 3. Ensure all the CI tests can pass.
> 
> @wyli Could you please also help take a look? Do you have any other comments?
> 
> Thanks.

Thank you Nic.
1. I have added `Enum` and `convert_to_list` to:
  https://github.com/Project-MONAI/MONAI/blob/dev/monai/utils/__init__.py
2. I added `convert_box_mode` and `convert_box_to_standard_mode` to docs. I was wondering if there are more I need to add. Do I need to add the `BoxMode` classes in `box_mode.py`?
3. The `mypy` test in `flake8` could not pass. How did you handle this case previously? Do you enforce `mypy` to pass?

Nic-Ma commented:
Hi @Can-Zhao ,

Please add `BoxMode` to the docs.

Thanks.

Nic-Ma commented:
/black

Can-Zhao commented:
> Hi @Can-Zhao ,
> 
> Please add `BoxMode` to the docs.
> 
> Thanks.

Done. Thanks!

wyli commented on monai/utils/type_conversion.py:
```diff
@@ -301,3 +301,15 @@ def convert_to_dst_type(
     return convert_data_type(data=src, output_type=output_type, device=device, dtype=dtype, wrap_sequence=wrap_sequence)
+
+
+def convert_to_list(data: Union[Sequence, torch.Tensor, np.ndarray]) -> list:
```
is a list really needed here? perhaps this could be done with the existing `ensure_tuple` utility?
```py
>>> from monai.utils import ensure_tuple
>>> ensure_tuple(torch.zeros((3, 4)))
(tensor([0., 0., 0., 0.]), tensor([0., 0., 0., 0.]), tensor([0., 0., 0., 0.]))
>>> import numpy as np
>>> ensure_tuple(np.zeros((3, 4)))                                                                                                                                
(array([0., 0., 0., 0.]), array([0., 0., 0., 0.]), array([0., 0., 0., 0.]))
```

Can-Zhao replied:
> is a list really needed here? perhaps this could be done with the existing `ensure_tuple` utility?
> 
> ```python
> >>> from monai.utils import ensure_tuple
> >>> ensure_tuple(torch.zeros((3, 4)))
> (tensor([0., 0., 0., 0.]), tensor([0., 0., 0., 0.]), tensor([0., 0., 0., 0.]))
> >>> import numpy as np
> >>> ensure_tuple(np.zeros((3, 4)))                                                                                                                                
> (array([0., 0., 0., 0.]), array([0., 0., 0., 0.]), array([0., 0., 0., 0.]))
> ```
I tried and got
```
>>> ensure_tuple(torch.zeros((3, )))
(tensor(0.), tensor(0.), tensor(0.))
```
I would expect 
```
>>> convert_to_list(torch.zeros((3, )))
[0., 0., 0.]
```
May be it's OK to not use it. But I need to test the code to know

wyli replied:
sure, I'll merge this PR for now but please check, intuitively the module should also support `(tensor(0.), tensor(0.), tensor(0.))`, it'll be useful for the refactoring for metatensor

wyli commented on monai/data/box_utils.py:
```diff
@@ -0,0 +1,565 @@
+    An abstract class of a ``BoxMode``.
+    A BoxMode is callable that converts box mode of boxes.
+    It always creates a copy and will not modify boxes in place.
+
```
I think it'll be great to put some summary of MONAI's bounding box mode in the docstring (either at the beginning of this module or in this class).

For example, we should highlight that this utility module mainly supports rectangular bounding boxes with a few different parameterizations and methods for converting between them. It provides reliable access to the spatial coordinates of the box vertices in the "canonical ordering" -- (x_start, y_start, x_end, y_end) for 2D and (x_start, y_start, z_start, x_end, y_end, z_end) for 3D.  We currently define this ordering as `monai.data.box_utils.StandardMode` and the rest of the detection pipelines mainly assumes boxes in `StandardMode`.

wyli commented on monai/data/box_utils.py:
```diff
@@ -0,0 +1,565 @@
+        src_mode: source box mode. If it is not given, this func will assume it is ``StandardMode``.
+        dst_mode: target box mode. If it is not given, this func will assume it is ``StandardMode``.
+
+    Note:
```
this note is duplicated, please put 
```
See also: :py:func:`monai.data.box_utils.get_boxmode`
```

wyli commented on monai/data/box_utils.py:
```diff
@@ -0,0 +1,565 @@
+    spatial_dims = get_spatial_dims(boxes=boxes_t)
+    for axis in range(0, spatial_dims):
+        if (corners[spatial_dims + axis] < corners[axis]).sum() > 0:
+            raise ValueError("Given boxes has invalid values. The box size must be non-negative.")
```
```suggestion
            warnings.warn("Given boxes have invalid values. The box size must be non-negative.")
```
perhaps this could be a warning for now? maybe it'll be useful in some applications?

wyli commented on monai/data/box_utils.py:
```diff
@@ -0,0 +1,565 @@
+        spatial_dims = int(spatial_dims_list[0])
+        spatial_dims = look_up_option(spatial_dims, supported=[2, 3])
+        return int(spatial_dims)
+    else:
+        raise ValueError("The dimensions of multiple inputs should match with each other.")
```
```suggestion
    raise ValueError("The dimensions of multiple inputs should match with each other.")
```

wyli reviewed (commented):
thanks, I put some docstring update suggestions

wyli commented:
/black

wyli commented:
/build

wyli commented:
/build
